Modules/_abc.c

```diff
@@ -800,32 +850,43 @@ _abc__abc_subclasscheck_impl(PyObject *module, PyObject *self,
         goto end;
     }
 
-    /* 6. Check if it's a subclass of a subclass (recursive). */
-    subclasses = PyObject_CallMethod(self, "__subclasses__", NULL);
-    if (subclasses == NULL) {
-        goto end;
-    }
-    if (!PyList_Check(subclasses)) {
-        PyErr_SetString(PyExc_TypeError, "__subclasses__() must return a list");
-        goto end;
-    }
-    for (pos = 0; pos < PyList_GET_SIZE(subclasses); pos++) {
-        PyObject *scls = PyList_GetItemRef(subclasses, pos);
-        if (scls == NULL) {
-            goto end;
-        }
-        int r = PyObject_IsSubclass(subclass, scls);
-        Py_DECREF(scls);
-        if (r > 0) {
-            if (_add_to_weak_set(impl, &impl->_abc_cache, subclass) < 0) {
-                goto end;
-            }
-            result = Py_True;
-            goto end;
-        }
-        if (r < 0) {
-            goto end;
-        }
+    /* 6. Check if it's a subclass of a subclass (recursive).
+        >>> class Ancestor: __subclasses__ = lambda: [Other]
+        >>> class Other: pass
+        >>> isinstance(Other, Ancestor) is True
+
+      Do not iterate over cls.__subclasses__() because it returns the entire class tree,
+      not just direct children, which leads to O(n^2) lookup.
+    */
+    PyObject *dict = _PyType_GetDict(cls); // borrowed
+    PyObject *subclasses_own_method = PyDict_GetItemString(dict, "__subclasses__"); // borrowed
+    if (subclasses_own_method) {
+      subclasses = PyObject_CallNoArgs(subclasses_own_method);
+      if (subclasses == NULL) {
+          goto end;
+      }
+      if (!PyList_Check(subclasses)) {
+          PyErr_SetString(PyExc_TypeError, "__subclasses__() must return a list");
+          goto end;
+      }
+      for (pos = 0; pos < PyList_GET_SIZE(subclasses); pos++) {
+          PyObject *scls = PyList_GetItemRef(subclasses, pos);
+          if (scls == NULL) {
+              goto end;
+          }
+          int r = PyObject_IsSubclass(subclass, scls);
```

Comment:
I think we have a UAF here. `PyObject_IsSubclass` can call `__subclasscheck__` which can itseslf call arbitrary code so you might mutate `subclasses`. The issue already exists with the existing code but can you confirm that we can indeed produce a UAF? (if you don't know how to do it, I'll try to investigate this separately tomorrow)